File: src/main/java/top/kmar/mi/content/tileentity/user/EUElectronSynthesizer.java
package top.kmar.mi.content.tileentity.user;

import net.minecraft.block.state.IBlockState;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.ITickable;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.items.ItemStackHandler;
import top.kmar.mi.api.araw.interfaces.AutoSave;
import top.kmar.mi.api.craftguide.ItemElement;
import top.kmar.mi.api.craftguide.sol.ItemList;
import top.kmar.mi.api.craftguide.sol.ItemSet;
import top.kmar.mi.api.electricity.clock.OrdinaryCounter;
import top.kmar.mi.api.electricity.info.BiggerVoltage;
import top.kmar.mi.api.electricity.info.EleEnergy;
import top.kmar.mi.api.electricity.info.EnumBiggerVoltage;
import top.kmar.mi.api.graphics.GuiLoader;
import top.kmar.mi.api.graphics.components.ProgressBarCmpt;
import top.kmar.mi.api.register.block.annotations.AutoTileEntity;
import top.kmar.mi.api.tools.FrontTileEntity;
import top.kmar.mi.api.utils.ItemUtil;
import top.kmar.mi.api.utils.WorldUtil;
import top.kmar.mi.content.blocks.BlockGuiList;
import top.kmar.mi.content.blocks.CraftList;
import top.kmar.mi.data.properties.MIProperty;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.List;

/**
 * 电子工作台
 * @author EmptyDreams
 */
@AutoTileEntity("electron_synthesizer")
@Mod.EventBusSubscriber
public class EUElectronSynthesizer extends FrontTileEntity implements ITickable {
	
	public static final int VOLTAGE = EleEnergy.COMMON;
	
	@AutoSave
    private final ItemStackHandler items = new ItemStackHandler(5 * 5 + 4);
	/** 工作时间 */
	@AutoSave private int workingTime = -10;
	@AutoSave private int maxTime = 0;
	@AutoSave private final List<ItemElement> OUTPUT = new ArrayList<>(4);
	@AutoSave private List<ItemStack> MERGE;
	/** 是否需要重新计算，当输入栏变动时需要将此项设置为true */
	private volatile boolean refresh = false;
	
	public EUElectronSynthesizer() {
		OrdinaryCounter counter = new OrdinaryCounter(100);
		counter.setBigger(new BiggerVoltage(2F, EnumBiggerVoltage.BOOM));
		setCounter(counter);
		setMaxEnergy(20);
	}
	
	@Override
	public void update() {
		if (world.isRemote) {
			WorldUtil.removeTickable(this);
			return;
		}
		if (refresh) {
			refresh = false;
			ItemList input = calculateProduction();
			ItemSet output = CraftList.SYNTHESIZER.apply(input.offset());
			if (check(output)) {
				workingTime = Math.max(workingTime, 0);
				maxTime = getMaxTime(input, output);
				IBlockState newState = world.getBlockState(
						getPos()).withProperty(MIProperty.getWORKING(), true);
				WorldUtil.setBlockState(world, pos, newState);
			} else {
				clear();
			}
		}
		if (!OUTPUT.isEmpty() && shrinkEnergy(1) && ++workingTime > maxTime) {
			export();
			clear();
		}
	}
	
	/** 清除状态 */
	private void clear() {
		OUTPUT.clear();
		workingTime = -2;
		MERGE = null;
		refresh = true;
		IBlockState newState = world.getBlockState(
				getPos()).withProperty(MIProperty.getWORKING(), false);
		WorldUtil.setBlockState(world, pos, newState);
	}
	
	/** 输出产物 */
	private void export() {
		for (int i = 0; i < 25; ++i) {
			items.getStackInSlot(i).shrink(1);
		}
		for (int i = 0; i < 4; ++i) {
			if (i < MERGE.size()) {
				items.setStackInSlot(i + 25, MERGE.get(i));
			} else {
				items.setStackInSlot(i + 25, ItemStack.EMPTY);
			}
		}
	}
	
	/** 检查输出是否可用以及是否与已有物品列表冲突 */
	private boolean check(ItemSet output) {
		if (output == null) {
			workingTime = -10;
			return false;
		}
		OUTPUT.clear();
		output.forEach(OUTPUT::add);
		merge();
		if (MERGE == null) {
			workingTime = -10;
			return false;
		}
		return true;
	}
	
	/** 根据原料列表计算产物 */
	private ItemList calculateProduction() {
		ItemList input = new ItemList(5, 5);
		for (int y = 0; y < 5; ++y) {
			for (int x = 0; x < 5; ++x) {
				input.set(x, y, ItemElement.instance(items.getStackInSlot(5 * y + x)));
			}
		}
		return input;
	}
	
	/** 合并产物列表 */
	private void merge() {
		ItemStack[] stack = new ItemStack[8];
		for (int i = 0; i < OUTPUT.size(); i++) {
			stack[i] = OUTPUT.get(i).getStack();
			stack[i + 4] = items.getStackInSlot(i + 25);
		}
		List<ItemStack> merge = ItemUtil.merge(stack);
		if (merge.size() < 4) MERGE = merge;
	}
	
	/** 获取工作需要的时间 */
	@SuppressWarnings("unused")
	public int getMaxTime(ItemList input, ItemSet output) {
		return 100;
	}
	
	@Nullable
	@Override
	public EnumFacing getFront() {
		return EnumFacing.UP;
	}
	
	@Override
	public boolean onReceive(EleEnergy energy) {
		if (energy.getVoltage() > VOLTAGE) getCounter().plus();
		return true;
	}
	
	@Override
	public boolean isReceiveAllowable(EnumFacing facing) {
		return true;
	}
	
	@Override
	public boolean isExtractAllowable(EnumFacing facing) {
		return false;
	}
	
	@Override
	public int getExVoltage() {
		return EleEnergy.COMMON;
	}
	
	public List<ItemStack> getAllStacks() {
		List<ItemStack> result = new ArrayList<>(29);
		for (int i = 0; i != 29; ++i) {
			result.add(items.getStackInSlot(i));
		}
		return result;
	}
	
	@SuppressWarnings("ConstantConditions")
	@SubscribeEvent
	public static void initGui(GuiLoader.MIGuiRegistryEvent event) {
		event.registryInitTask(BlockGuiList.getSynthesizer(), gui -> {
			EUElectronSynthesizer synthesizer = (EUElectronSynthesizer) gui.getTileEntity();
			gui.initItemStackHandler(synthesizer.items);
		});
		event.registryLoopTask(BlockGuiList.getSynthesizer(), gui -> {
			EUElectronSynthesizer synthesizer = (EUElectronSynthesizer) gui.getTileEntity();
			ProgressBarCmpt work = (ProgressBarCmpt) gui.getElementByID("work");
			work.setMax(synthesizer.maxTime);
			work.setValue(synthesizer.workingTime);
		});
	}
	
}

File: src/main/java/top/kmar/mi/content/tileentity/user/EUElectronSynthesizer.kt
package top.kmar.mi.content.tileentity.user

import net.minecraft.item.ItemStack
import net.minecraft.util.EnumFacing
import net.minecraft.util.ITickable
import net.minecraftforge.fml.common.Mod.EventBusSubscriber
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.items.ItemStackHandler
import top.kmar.mi.api.araw.interfaces.AutoSave
import top.kmar.mi.api.craft.CraftGuide
import top.kmar.mi.api.craft.elements.CraftOutput
import top.kmar.mi.api.craft.elements.ElementList
import top.kmar.mi.api.electricity.EleEnergy
import top.kmar.mi.api.graphics.BaseGraphics
import top.kmar.mi.api.graphics.GuiLoader.MIGuiRegistryEvent
import top.kmar.mi.api.graphics.components.ProgressBarCmpt
import top.kmar.mi.api.graphics.components.SlotMatrixCmpt
import top.kmar.mi.api.regedits.block.annotations.AutoTileEntity
import top.kmar.mi.api.tools.FrontTileEntity
import top.kmar.mi.api.utils.*
import top.kmar.mi.api.utils.expands.*
import top.kmar.mi.content.blocks.BlockGuiList.synthesizer
import top.kmar.mi.data.CraftList
import top.kmar.mi.data.properties.MIProperty.Companion.WORKING
import kotlin.math.min

/**
 * 电子工作台
 * @author EmptyDreams
 */
@AutoTileEntity("electron_synthesizer")
@EventBusSubscriber
class EUElectronSynthesizer : FrontTileEntity(), ITickable {

    /** 物品列表 */
    @field:AutoSave
    private val items = ItemStackHandler(5 * 5 + 4)
    /** 工作时间 */
    @field:AutoSave
    private var workProgress = 0
    /** 下一次的输出 */
    private var output: CraftOutput? = null
    /** 获取一个部件加工完成所需的能量 */
    val needEnergy: Int
        get() = output!!.getInt("energy", 20000)

    override fun update() {
        if (world.isClient()) return removeTickable()
        if (output == null) {
            this.output = calculateOutput()
            if (!putOutput(true)) {
                removeTickable()
                return clear(false)
            }
        }
        val need = needEnergy
        val energy = requestEnergy(min(100, need))
        if (energy.isEmpty) return updateBlockState(false)
        if (energy.voltage < minVoltage) return updateBlockState(true)
        if (energy.voltage > maxVoltage) return explode(3.5F, true)
        workProgress += energy.capacity
        if (workProgress >= need) {
            !putOutput(false)
            clear(false)
        } else updateBlockState(true)
        markDirty()
    }

    /** 计算输出，返回`null`表示无输出 */
    fun calculateOutput(): CraftOutput? {
        val existing = getOutputStacks()
        if (existing.all { it.count >= it.maxStackSize }) return null
        val input = calculateProduction()
        val output = CraftGuide.findOutput(CraftList.synthesizer, input)
        if (output == null) clear(true)
        else {
            val surplus = items.insertItems(25 until 29, output.stacks, true)
            if (surplus.isNotEmpty()) return null
        }
        return output
    }

    /**
     * 将输出添加到物品框中
     * @param simulation 是否为模拟
     * @return 是否成功放入
     */
    fun putOutput(simulation: Boolean): Boolean {
        val output = this.output?.stacks ?: return false
        val itor = output.iterator()
        var value: ItemStack = ItemStack.EMPTY
        for (i in 25 until 29) {
            if (value.isEmpty) {
                if (!itor.hasNext()) break
                value = itor.next()
            }
            val insert = items.insertItem(i, value.copy(), simulation)
            value.count = insert.count
        }
        if (itor.hasNext()) return false
        if (!simulation) {
            for (i in 0 until 25) {
                val stack = items.getStackInSlot(i)
                if (!stack.isEmpty) stack.shrink(1)
            }
        }
        return true
    }

    /** 根据原料列表计算产物  */
    fun calculateProduction(): ElementList {
        val input = ElementList(5, 5)
        for (y in 0..4) {
            for (x in 0..4) {
                input[x, y] = items.getStackInSlot(5 * y + x).copy()
            }
        }
        return input
    }

    /** 清楚工作状态和进度 */
    fun clear(updateState: Boolean) {
        workProgress = 0
        output = null
        if (updateState) updateBlockState(false)
    }

    /** 更新方块显示状态 */
    fun updateBlockState(isWorking: Boolean) {
        val old = world.getBlockState(pos)
        if (old.getValue(WORKING) == isWorking) return
        val newState = old.withProperty(WORKING, isWorking)
        world.setBlockWithMark(pos, newState)
    }

    override fun getFront() = EnumFacing.UP

    /** 获取所有输出框的stack */
    fun getOutputStacks() = Array(4) { items.getStackInSlot(25 + it) }

    /** 获取所有物品框的stack */
    fun getAllStacks(): List<ItemStack> {
        val result = ArrayList<ItemStack>(items.slots)
        for (i in 0 until items.slots) {
            result.add(items.getStackInSlot(i))
        }
        return result
    }

    companion object {

        /** 电器可以承受的最大电压 */
        const val maxVoltage = EleEnergy.COMMON + 50
        /** 电器可承受的最小电压 */
        const val minVoltage = EleEnergy.COMMON - 50

        @SubscribeEvent
        @JvmStatic
        fun initGui(event: MIGuiRegistryEvent) {
            event.registryInitTask(synthesizer) { gui: BaseGraphics ->
                val synthesizer =
                    gui.tileEntity as EUElectronSynthesizer
                gui.initItemStackHandler(synthesizer.items)
                val input = gui.getElementByID("input") as SlotMatrixCmpt
                input.slotAttributes.onSlotChanged = {
                    with(synthesizer) {
                        output = calculateOutput()
                        addTickable()
                    }
                }
                val output = gui.getElementByID("output") as SlotMatrixCmpt
                output.slotAttributes.onSlotChanged = { synthesizer.addTickable() }
            }
            event.registryLoopTask(synthesizer) { gui: BaseGraphics ->
                val synthesizer = gui.tileEntity as EUElectronSynthesizer
                val work = gui.getElementByID("work") as ProgressBarCmpt
                if (synthesizer.output == null) {
                    work.value = 0
                    work.max = 0
                } else {
                    work.max = synthesizer.needEnergy
                    work.value = synthesizer.workProgress
                }
            }
        }

    }

}